scwx-qt/source/scwx/qt/manager/marker_manager.cpp

```diff
@@ -37,12 +37,31 @@ class MarkerManager::Impl
    ~Impl() { threadPool_.join(); }
 
    std::string                                markerSettingsPath_ {""};
-   std::vector<std::shared_ptr<MarkerRecord>> markerRecords_ {};
 
    MarkerManager* self_;
 
    boost::asio::thread_pool threadPool_ {1u};
-   std::shared_mutex        markerRecordLock_ {};
+   /* Accessing markerRecords_
+    * For reads to markerRecords_, simply use std::shared_lock to lock
+    * markerRecordAccessMutex_
+    * For writes to markerRecords_
+    * 1. lock markerRecordModifyMutex_
+    * 2. emit start signals
+    * 3. lock (std::unique_lock) markerRecordAccessMutex_
+    * 4. Do changes
+    * 5. Unlock markerRecordAccessMutex_
+    * 6. emit end signals
```

Comment:
Slot connections (auto, which is the correct usage here) are not guaranteed to be called at the time of the signal being emitted, only for direct connections (when the slot object is in the same thread as the signal being emitted). When two the slot object is in a different thread (which happens on initialization), a queued connection is made, and the slot will be called in its own thread (which will occur during the next iteration of the Qt event loop).

I don't see any deadlocks, so this is probably OK.  I don't see any slots calling a function to modify.  Although, I think your error handling in the MarkerModel is sufficient to cover race conditions (if the index doesn't exist anymore, return empty QVariant, etc.).

Reply:
> Slot connections (auto, which is the correct usage here) are not guaranteed to be called at the time of the signal being emitted, only for direct connections (when the slot object is in the same thread as the signal being emitted). When two the slot object is in a different thread (which happens on initialization), a queued connection is made, and the slot will be called in its own thread (which will occur during the next iteration of the Qt event loop).

It seems to be working, but I was getting asserts within `beginRemoveRows` because I was removing from the manager before calling it. I think that could still happen then. I can think of two ways to fix it.
1. Turn the model into a singleton, and call functions directly (instead of using slots) from the manager.
2. Have all insert and delete calls go through the model first, then call the manager. This would likely require the model be a singleton as well.

1 will require fewer changes, but I think 2 is the closer to the way Qt intended models to work. I prefer 1, but I would not mind doing 2 either.

Reply:
I'm not sure I like having the manager call the model directly, or having all insert/delete go through the model (unless it's a combined model/manager similar to the layer model), I think the dependency should be the other way around. The way current managers/models are patterned in the code, the manager is the source of truth for the data, which I'd like to keep.

I think the model being a singleton is just fine. There are other singleton models already.

The placefile model and manager are probably the best examples of this working together currently. The placefile model keeps a thin copy of placefile names to use as a reference into the manager for other operations. To mitigate a potential data race, if the data in the underlying manager changes and the name goes stale, the placefile manager returns a default value.

Reply:
I have a branch that works more like placefile manager/model. I use a 64 bit integer as an ID. The ID's are local to the process, not saved to the json. Making IDs persistent would require more changes, but would be possible. It also would reverts the last two commits. I can integrate it into this pull request if you think this makes more sense then relying on things happening in the same thread for correct ordering.

Reply:
I can take a closer look later, but I don't think IDs necessarily need to be persistent.

Reply:
Because of how it diverged, I thought it would be better and simpler to have it be its own pull request.